library/cpp/include/wavemap/core/utils/profile/resource_monitor.h

```diff
@@ -0,0 +1,150 @@
+#ifndef WAVEMAP_CORE_UTILS_PROFILE_RESOURCE_MONITOR_H_
+#define WAVEMAP_CORE_UTILS_PROFILE_RESOURCE_MONITOR_H_
+
+#include <ctime>
+#include <fstream>
+#include <optional>
+#include <sstream>
+
+#include "wavemap/core/common.h"
+#include "wavemap/core/utils/time/time.h"
+
+namespace wavemap {
+/**
+ * @brief Monitors system resource usage, including CPU time, wall time,
+ *        and RAM usage.
+ *
+ * The `ResourceMonitor` class tracks CPU and wall clock time over timed
+ * episodes, much like wavemap's Stopwatch class. It also provides functionality
+ * to retrieve the total RAM usage of the current process.
+ */
+class ResourceMonitor {
+ public:
+  /**
+   * @brief Starts a new CPU and wall time monitoring episode.
+   *
+   * Records the CPU and wall clock start times for the current episode.
+   * If monitoring is already running, calling `start()` has no effect.
+   */
+  void start();
+
+  /**
+   * @brief Stops timing the current episode.
+   *
+   * Records the end CPU and wall clock times for the current episode,
+   * updating the last episode duration and total accumulated duration.
+   * If monitoring is not running, calling `stop()` has no effect.
+   */
+  void stop();
+
+  /**
+   * @brief Checks if the stopwatch is currently running.
+   *
+   * @return `true` if the stopwatch is running, `false` otherwise.
+   */
+  bool isRunning() const { return running_; }
+
+  /**
+   * @brief Gets the CPU time duration of the last episode.
+   *
+   * @return The CPU time duration of the last episode in seconds.
+   *
+   * The value represents the CPU time elapsed between the most recent
+   * `start()` and `stop()` calls. If no episode has been monitored, this
+   * returns 0.
+   */
+  double getLastEpisodeCpuTime() const {
```

Comment:
What is the intended behaviour of this when it is called while the monitor is active?

Reply:
It will return the CPU time from the last completed episode, or 0 if there was none.

Reply:
Is that what you'd intuitively expect?

Reply:
I updated the comments throughout the Stopwatch and ResourceMonitor class to communicate this behavior more clearly.